runtime/cudaq/operators/boson_op.cpp

```diff
@@ -149,44 +169,119 @@ boson_handler::boson_handler(std::size_t target, int op_id)
-      auto column = row - this->additional_terms;
-      mat[{row, column}] = 1.;
-      for (auto offset : this->number_offsets)
-        mat[{row, column}] *= (row + offset);
-      for (auto offset = -this->additional_terms; offset > 0; --offset)
-        mat[{row, column}] *= std::sqrt(row + offset);
+    tokens.push_back(s.substr(start));
+    return tokens;
+  };
+
+  auto map_state = [&tokenize](std::string encoding, std::int64_t old_state) {
+    if (encoding == "I")
+      return std::pair<double, std::int64_t>{1., old_state};
+    auto ops = tokenize(encoding, '.');
+    assert(ops.size() > 0);
+
+    auto it = ops.cbegin();
+    int additional_terms = std::stol(*it);
+    std::int64_t new_state = old_state + additional_terms;
+    if (new_state < 0)
+      return std::pair<double, std::int64_t>{0., old_state};
+
+    double value = 1.;
+    if (additional_terms > 0)
+      for (auto offset = additional_terms; offset > 0; --offset)
+        value *= std::sqrt(old_state + offset);
+    if (additional_terms < 0)
+      for (auto offset = -additional_terms; offset > 0; --offset)
+        value *= std::sqrt(new_state + offset);
+    while (++it != ops.cend())
+      value *= (new_state + std::stol(*it));
+    return std::pair<double, std::int64_t>{value, new_state};
+  };
+
+  auto states = cudaq::detail::generate_all_states(dimensions);
+  if (states.size() == 0)
+    process_element(0, 0, 1.);
+  std::vector<std::string> boson_terms = tokenize(boson_word, '_');
+  std::size_t old_state_idx = 0;
+  for (const auto &old_state : states) {
+    std::vector<std::int64_t> new_state(old_state.size(), 0);
+    std::complex<double> entry = 1.;
+    for (std::size_t degree = 0; degree < old_state.size(); ++degree) {
+      auto state = old_state[degree];
+      auto op =
+          boson_terms[invert_order ? old_state.size() - 1 - degree : degree];
+      auto mapped = map_state(op, state);
+      entry *= mapped.first;
+      if (mapped.second >= dimensions[degree])
+        entry = 0.;
+      else
+        new_state[degree] = mapped.second;
     }
-  } else {
-    for (std::size_t i = 0; i < dim; i++) {
-      mat[{i, i}] = 1.;
-      for (auto offset : this->number_offsets)
-        mat[{i, i}] *= (i + offset);
+
+    if (entry != 0.) {
+      auto new_state_idx = 0;
+      for (std::size_t idx = 0; idx < new_state.size(); ++idx) {
+        auto offset = 1;
+        for (std::size_t d = 0; d < idx; ++d)
+          offset *= dimensions[d];
+        new_state_idx += new_state[idx] * offset;
+      }
+      process_element(new_state_idx, old_state_idx, entry);
     }
+    old_state_idx += 1;
   }
-  return mat;
+}
+
+cudaq::detail::EigenSparseMatrix
+boson_handler::to_sparse_matrix(const std::string &boson_word,
+                                const std::vector<std::int64_t> &dimensions,
+                                std::complex<double> coeff, bool invert_order) {
+  auto dim = 1;
```

Comment:
nit: we might want to be explicit about the type here (`std::int64_t`) to prevent overflow if a 32-bit integer is deducted.